File: examples/C/fileio3.c
//  File Transfer model #3
//  
//  In which the client requests each chunk individually, using
//  command pipelining to give us a credit-based flow control.

#include <czmq.h>
#define CHUNK_SIZE  250000
#define PIPELINE    10

static void
client_thread (void *args, zctx_t *ctx, void *pipe)
{
    void *dealer = zsocket_new (ctx, ZMQ_DEALER);
    zsocket_set_hwm (dealer, PIPELINE);
    zsocket_connect (dealer, "tcp://127.0.0.1:6000");

    //  We'll allow up to five chunks in transit at once
    size_t credit = PIPELINE;
    
    size_t total = 0;       //  Total bytes received
    size_t chunks = 0;      //  Total chunks received
    size_t offset = 0;      //  Offset of next chunk request
    
    while (true) {
        while (credit) {
            //  Ask for next chunk
            zstr_sendfm (dealer, "fetch");
            zstr_sendfm (dealer, "%ld", offset);
            zstr_sendf  (dealer, "%ld", CHUNK_SIZE);
            offset += CHUNK_SIZE;
            credit--;
        }
        zframe_t *chunk = zframe_recv (dealer);
        if (!chunk)
            break;              //  Shutting down, quit
        chunks++;
        credit++;
        size_t size = zframe_size (chunk);
        zframe_destroy (&chunk);
        total += size;
        if (size < CHUNK_SIZE)
            break;              //  Last chunk received; exit
    }
    printf ("%zd chunks received, %zd bytes\n", chunks, total);
    zstr_send (pipe, "OK");
}

//  The rest of the code is exactly the same as in model 2, except
//  that we set the HWM on the server's ROUTER socket to PIPELINE
//  to act as a sanity check.
//  .skip
static void
free_chunk (void *data, void *arg)
{
    free (data);
}

//  The server thread waits for a chunk request from a client,
//  reads that chunk and sends it back to the client:

static void
server_thread (void *args, zctx_t *ctx, void *pipe)
{
    FILE *file = fopen ("testdata", "r");
    assert (file);

    void *router = zsocket_new (ctx, ZMQ_ROUTER);
    zsocket_set_hwm (router, PIPELINE);
    zsocket_bind (router, "tcp://*:6000");
    while (true) {
        //  First frame in each message is the sender identity
        zframe_t *identity = zframe_recv (router);
        if (!identity)
            break;              //  Shutting down, quit
            
        //  Second frame is "fetch" command
        char *command = zstr_recv (router);
        assert (streq (command, "fetch"));
        free (command);

        //  Third frame is chunk offset in file
        char *offset_str = zstr_recv (router);
        size_t offset = atoi (offset_str);
        free (offset_str);

        //  Fourth frame is maximum chunk size
        char *chunksz_str = zstr_recv (router);
        size_t chunksz = atoi (chunksz_str);
        free (chunksz_str);

        //  Read chunk of data from file
        fseek (file, offset, SEEK_SET);
        byte *data = malloc (chunksz);
        assert (data);

        //  Send resulting chunk to client
        size_t size = fread (data, 1, chunksz, file);
        zframe_t *chunk = zframe_new_zero_copy (data, size, free_chunk, NULL);
        zframe_send (&identity, router, ZFRAME_MORE);
        zframe_send (&chunk, router, 0);
    }
    fclose (file);
}


//  The main task starts the client and server threads; it's easier
//  to test this as a single process with threads, than as multiple
//  processes:

int main (void)
{
    //  Start child threads
    zctx_t *ctx = zctx_new ();
    zthread_fork (ctx, server_thread, NULL);
    void *client =
    zthread_fork (ctx, client_thread, NULL);
    //  Loop until client tells us it's done
    char *string = zstr_recv (client);
    free (string);
    //  Kill server thread
    zctx_destroy (&ctx);
    return 0;
}
//  .until

File: examples/Java/Fileio3.java
import org.zeromq.ZContext;
import org.zeromq.ZFrame;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZThread;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Arrays;

//  File Transfer model #3
//
//  In which the client requests each chunk individually, using
//  command pipelining to give us a credit-based flow control.
public class Fileio3 {
  private static final int PIPELINE = 10;
  private static final int CHUNK_SIZE = 250000;

  //  The main task starts the client and server threads; it's easier
  //  to test this as a single process with threads, than as multiple
  //  processes:
  public static void main(String[] args) {
    ZContext ctx = new ZContext();
    //  Start child threads
    ZThread.fork(ctx, new Fileio1.Server());
    ZMQ.Socket client = ZThread.fork(ctx, new Fileio1.Client());
    //  Loop until client tells us it's done
    client.recvStr();
    //  Kill server thread
    ctx.destroy();
  }

  static class Client implements ZThread.IAttachedRunnable {

    @Override
    public void run(Object[] args, ZContext ctx, Socket pipe) {
      Socket dealer = ctx.createSocket(ZMQ.DEALER);
      dealer.connect("tcp://127.0.0.1:6000");

      //  Up to this many chunks in transit
      int credit = PIPELINE;

      int total = 0;       //  Total bytes received
      int chunks = 0;      //  Total chunks received
      int offset = 0;      //  Offset of next chunk request

      while (true) {
        while (credit > 0) {
          //  Ask for next chunk
          dealer.sendMore("fetch");
          dealer.sendMore(String.valueOf(offset));
          dealer.send(String.valueOf(CHUNK_SIZE));
          offset += CHUNK_SIZE;
          credit--;
        }
        ZFrame chunk = ZFrame.recvFrame(dealer);
        if (chunk.getData() == null)
          break;    //  Shutting down, quit
        chunks++;
        credit++;
        int size = chunk.size();
        chunk.destroy();
        total += size;
        if (size < CHUNK_SIZE)
          break;              //  Last chunk received; exit
      }
      System.out.printf("%d chunks received, %d bytes\n", chunks, total);
      pipe.send("OK");
    }
  }
  //  The rest of the code is exactly the same as in model 2, except
//  that we set the HWM on the server's ROUTER socket to PIPELINE
//  to act as a sanity check.

  //  The server thread waits for a chunk request from a client,
//  reads that chunk and sends it back to the client:
  static class Server implements ZThread.IAttachedRunnable {

    @Override
    public void run(Object[] args, ZContext ctx, Socket pipe) {
      File file = new File("testdata");
      FileInputStream fr;
      try {
        fr = new FileInputStream(file);
      } catch (FileNotFoundException e) {
        e.printStackTrace();
        return;
      }
      ZMQ.Socket router = ctx.createSocket(ZMQ.ROUTER);
      router.setHWM(PIPELINE * 2);
      router.bind("tcp://*:6000");
      while (!Thread.currentThread().isInterrupted()) {
        //  First frame in each message is the sender identity
        ZFrame identity = ZFrame.recvFrame(router);
        if (identity.getData() == null)
          break; //  Shutting down, quit

        //  Second frame is "fetch" command
        String command = router.recvStr();
        assert ("fetch".equals(command));

        //  Third frame is chunk offset in file
        int offset = Integer.parseInt(router.recvStr());

        //  Fourth frame is maximum chunk size
        int chunkSize = Integer.parseInt(router.recvStr());

        //  Read chunk of data from file
        byte[] data = new byte[CHUNK_SIZE];
        int size;
        try {
          fr.skip(offset);
          size = fr.read(data, 0, chunkSize);
        } catch (IOException e) {
          e.printStackTrace();
          break;
        }

        //  Send resulting chunk to client
        ZFrame chunk = new ZFrame(Arrays.copyOf(data, size < 0 ? 0 : size));
        identity.sendAndDestroy(router, ZMQ.SNDMORE);
        chunk.sendAndDestroy(router);
      }
      try {
        fr.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
  }
}

